workflow-core/src/commonMain/kotlin/com/squareup/workflow1/WorkflowAction.kt

```diff
@@ -133,3 +134,28 @@ public class WorkflowOutput<out OutputT>(public val value: OutputT) : ActionProc
 
   override fun hashCode(): Int = value.hashCode()
 }
+
+/**
+ * An [ActionProcessingResult] is any possible outcome after the runtime does a loop of processing.
+ *
+ * Only [ActionApplied] needs the generic OutputT so we do not include it in the root
+ * interface here.
+ */
+public sealed interface ActionProcessingResult
+
+public object PropsUpdated : ActionProcessingResult
+
+public object ActionsExhausted : ActionProcessingResult
+
+/**
+ * Result of applying an action.
+ *
+ * @param output: the potentially null [WorkflowOutput]. If null, then no output was set by the
+ *   action. Otherwise it is a [WorkflowOutput] around the output value of type [OutputT],
+ *   which could be null.
+ * @param stateChanged: whether or not the action changed the state.
+ */
+public data class ActionApplied<out OutputT>(
```

Comment:
Do we actually rely on this being a `data class`? I remember that being a landmine when used in public API, though I forget exactly why.

Reply:
Remember that you can implement positional params manually, it's pretty trivial.

Reply:
Like to support destructuring you mean?

Hmm, now I'm curious - what the was the data class issue in the public API?

Reply:
For posterity. I am going to do this change after all. Canonical resource describing backwards compatibility concerns: https://jakewharton.com/public-api-challenges-in-kotlin/